### PR TITLE
Simplify API for making archives.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs [hunk 1]
```diff
@@ -10,19 +10,18 @@ namespace Pulumi
 {
     public sealed partial class Deployment
     {
-        Task IDeployment.InvokeAsync(string token, ResourceArgs args, InvokeOptions? options)
+        Task IDeployment.InvokeAsync(string token, ResourceArgs? args, InvokeOptions? options)
```
Comment (mikhailshilkov) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]: I've been changing tfgen not to pass nulls here, ready to open a PR for that. Avoiding nulls sounds like a better decision to me?

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]: https://github.com/pulumi/pulumi-terraform-bridge/pull/69

Comment (CyrusNajmabadi) [reply sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]: i went back and forth on this (i made basically the exact same change to the bridge :)  But i ended updeciding i didn't like pushing this all the way out to every single caller. It seemed reasonable to me to say: you can pass me null and i'm ok with that here. 

I'm find with either approach.  one benefit of mine is that we dont have to update every resource provider.

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]: Agreed. I prefer avoiding defensive null checks, so I'd keep it non-nullable. Your workaround should stay for now until all providers are organically updated.

Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]:
```suggestion
        Task IDeployment.InvokeAsync(string token, ResourceArgs args, InvokeOptions? options)
```

File: sdk/dotnet/Pulumi/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -106,7 +106,7 @@ Pulumi.Output<T>.Apply<U>(System.Func<T, Pulumi.Output<U>> func) -> Pulumi.Outpu
 Pulumi.Output<T>.Apply<U>(System.Func<T, System.Threading.Tasks.Task<U>> func) -> Pulumi.Output<U>
 Pulumi.Output<T>.Apply<U>(System.Func<T, U> func) -> Pulumi.Output<U>
 Pulumi.ProviderResource
-Pulumi.ProviderResource.ProviderResource(string package, string name, Pulumi.ResourceArgs args, Pulumi.ResourceOptions options = null) -> void
+Pulumi.ProviderResource.ProviderResource(string package, string name, Pulumi.ResourceArgs args = null, Pulumi.ResourceOptions options = null) -> void
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/PublicAPI.Unshipped.txt]:
```suggestion
Pulumi.ProviderResource.ProviderResource(string package, string name, Pulumi.ResourceArgs args, Pulumi.ResourceOptions options = null) -> void
```

File: sdk/dotnet/Pulumi/Resources/ProviderResource.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public class ProviderResource : CustomResource
         /// </summary>
         public ProviderResource(
             string package, string name,
-            ResourceArgs args, ResourceOptions? options = null)
+            ResourceArgs? args = null, ResourceOptions? options = null)
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Resources/ProviderResource.cs]:
```suggestion
            ResourceArgs args, ResourceOptions? options = null)
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs [hunk 4]
```diff
@@ -13,16 +13,15 @@ public sealed partial class Deployment
...
 
         private async Task<T> InvokeAsync<T>(
-            string token, ResourceArgs args, InvokeOptions? options, bool convertResult)
+            string token, ResourceArgs? args, InvokeOptions? options, bool convertResult)
         {
             var label = $"Invoking function: token={token} asynchronously";
             Log.Debug(label);
 
-            // Be resilient to misbehaving callers.
             args ??= ResourceArgs.Empty;
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]:
```suggestion
            // Be resilient to misbehaving callers.
            args ??= ResourceArgs.Empty;
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs [hunk 5]
```diff
@@ -13,11 +13,11 @@ public sealed partial class Deployment
         Task IDeployment.InvokeAsync(string token, ResourceArgs args, InvokeOptions? options)
             => InvokeAsync<object>(token, args, options, convertResult: false);
 
-        Task<T> IDeployment.InvokeAsync<T>(string token, ResourceArgs args, InvokeOptions? options)
+        Task<T> IDeployment.InvokeAsync<T>(string token, ResourceArgs? args, InvokeOptions? options)
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]:
```suggestion
        Task<T> IDeployment.InvokeAsync<T>(string token, ResourceArgs args, InvokeOptions? options)
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ Task<T> IDeployment.InvokeAsync<T>(string token, ResourceArgs args, InvokeOption
             => InvokeAsync<T>(token, args, options, convertResult: true);
 
         private async Task<T> InvokeAsync<T>(
-            string token, ResourceArgs args, InvokeOptions? options, bool convertResult)
+            string token, ResourceArgs? args, InvokeOptions? options, bool convertResult)
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs]:
```suggestion
            string token, ResourceArgs args, InvokeOptions? options, bool convertResult)
```

**Pull request merged**